# Exploring Ebay Car Sales Data

The goal of this project is to clean and analyze car sales data in an exploratory manner. This information should provide value in the purchasing of a used car in the German market. The dataset is complied of used car data from eBay Kleinanzeigen (a classified section of the German eBay website). The dataset was originally scraped and [uploaded](https://data.world/data-society/used-cars-data) to Kaggle by user orgesleka who is no longer active.

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
os.getcwd()

'C:\\Users\\gfulham\\Dataquest_JPNBs\\2_data_analysis_visualisation_exercises'

## 1) Glance at Data

In [3]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [6]:
autos.isnull().sum()

dateCrawled                0
name                       0
seller                     0
offerType                  0
price                      0
abtest                     0
vehicleType            37869
yearOfRegistration         0
gearbox                20209
powerPS                    0
model                  20484
kilometer                  0
monthOfRegistration        0
fuelType               33386
brand                      0
notRepairedDamage      72060
dateCreated                0
nrOfPictures               0
postalCode                 0
lastSeen                   0
dtype: int64

### What stands out:
- There are 371528 entries, which is a large enough sample size
- The Name column's values are seperated by underscores, making it easy to slice.
- The Names of the columns are in camelcase, not the preferred snakecase. 
- Most dtypes are objects, but objects like dateCrawled, date created, price, odometer,  can be changed to be easier to work with. 
- notRepairDamage has almost 20% of its values missing. Something to deal with down the road. 

In [9]:
# Lets manualy change the column names from camelcase to snakecase. 
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [10]:
# Create new column names. Create a dictionary of names we need changed.{old,new}
# These changes will make it easier to work with these columns in the future. 
autos = autos.rename ({'dateCrawled': 'date_crawled',
                       'offerType': 'offer_type',
                       'vehicleType' : 'vehicle_type', 
                       'yearOfRegistration' : 'registration_year', 
                       'powerPS': 'power_ps', 
                       'monthOfRegistration': 'registration_month',
                       'fuelType': 'fuel_type', 
                       'notRepairedDamage': 'unrepaired_damage',
                       'dateCreated': 'ad_created',
                       'nrOfPictures': 'num_pictures',
                       'postalCode': 'postal_code',
                       'lastSeen': 'last_seen',
                       'kilometer': 'odometer_km',
                        }, axis= 1)

autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   0    benzin  volkswagen               NaN   
1       125000                   5    diesel        audi                ja   
2       125000                   8    diesel        jeep               NaN   
3       150000                   6    benzin  volkswagen              nein   
4        90000                   7    diesel       skoda              nein   

            ad_created  num_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00             0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0        60437  2016-04-06 10:17:21

## 2) Data cleaning and exploration
Look for:
- Text columns where all the data are the same.
- Columns with wrong dtype.

In [11]:
# Lets look at the descriptive statistics for all columns, objects and ints. 
autos.describe(include = 'all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model    odometer_km  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_pictures   postal_code  \
count             299468               371528      371528.0  371528.00000   
unique                 2                  114           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq              263182                14450           NaN           NaN   
mean                 NaN                  NaN           0.0   50820.66764   
std                  NaN                  NaN           0.0   25799.08247   
min                  NaN                  NaN           0.0    1067.00000   
25%                  NaN                  NaN           0.0   30459.00000   
50%                  NaN                  NaN           0.0   49610.00000   
75%                  NaN                  NaN           0.0   71546.00000   
max                  NaN                  NaN           0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### What stands out:
1. **num_picures** column has 0 for all the entries. The **seller** and **offer_type** columns each have 375500+ identical entries. Therefore all three should be dropped.
2. **registration_year** max is 9999. Values above 2016 should be NaN
3. **power_ps** max is 20000

In [12]:
# Removing seller, num_pictures, and offer type columns. Use df.drop()
autos = autos.drop(['seller', 'num_pictures', 'offer_type'], axis = 1)
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [13]:
# This has been commented out becuase the origional dataset does not need to be fixed.
# This code changed the values in the price and odometer columns. 

# autos['price'] = (
#     autos['price']
#     .str.replace('$','')
#     .str.replace(',','')
#     .astype(int)
# )

# autos['odometer'] = (
#     autos['odometer']
#     .str.replace('km','')
#     .str.replace(',','')
#     .astype(int)
# )

# # Rename the columns since they kn longer have $ and KM
# autos.rename(columns= {'price': 'price_$'}, inplace = True)    
# autos.rename(columns= {'odometer':'odometer_km'},  inplace = True)  

# autos.describe() # Check  

## 3) Examining Values in Odometer and Price
From looking at the descriptive statistics, the max price and max odometer values initially seem wrong. Theses columns need to be further examined. For examining the data to remove outliers, we will inspect columns odometer_km and price_$ seperatly. 

In [14]:
print(autos["odometer_km"].unique().shape, '\n')
print(autos["odometer_km"].describe(),'\n')
print(autos["odometer_km"].value_counts())

(13,) 

count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64


**Odometer observations**\
There are 13 different bins in which the odomoeter_km contains values. The min and max kms seem normal, and there is a pretty equal and expected distribution amoung the used cars. 

In [16]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().head()) # value counts will count the number of time a value appears in the price_$ column.

(5597,)
count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64
0       10778
500      5670
1500     5394
1000     4649
1200     4594
Name: price, dtype: int64


**Price Observations**\
The most expensive car is valued over $2,000,000,000 which is wrong. Lets further examine the top most expensive cars untill we find a resonable number to call our cut-off point. 

In [17]:
print(autos['price'].value_counts().sort_index().tail(60))# Here use sort_index to sort the values in the column itself. 
print(autos['price'].value_counts().sort_index().head(20))

294900         1
295000         1
299000         3
300000         1
323223         1
345000         1
349000         1
350000         4
368000         1
370000         2
390000         1
395000         1
399997         1
420000         1
440000         1
445000         1
466000         1
485000         1
487000         1
488997         1
500000         2
517895         1
579000         1
585000         1
599000         1
600000         2
619000         1
650000         1
700000         1
725000         1
745000         2
780000         1
820000         1
849000         1
911911         1
999990         1
999999        13
1000000        5
1010010        1
1111111        2
1234566        1
1250000        2
1300000        1
1600000        2
2795000        1
2995000        1
3890000        1
3895000        1
9999999        3
10000000       8
10010011       1
11111111      10
12345678       9
14000500       1
27322222       1
32545461       1
74185296       1
99000000       1
99999999      

In [18]:
#Examine some high priced vehicles to determine their viability. 
high_priced = autos.loc[autos['price'] == 999999]
high_priced

date_crawled  \
38445   2016-03-21 19:53:52   
69726   2016-03-17 15:58:57   
118233  2016-03-09 21:56:45   
131785  2016-04-03 18:58:58   
141004  2016-04-02 13:47:00   
155418  2016-03-27 20:52:14   
159974  2016-04-01 16:36:57   
169436  2016-04-01 12:56:01   
169656  2016-03-21 20:45:24   
174617  2016-03-14 21:40:52   
301198  2016-04-04 09:36:42   
312909  2016-03-17 09:53:08   
332531  2016-03-20 02:36:23   

                                                     name   price   abtest  \
38445                                       2_VW_Busse_T3  999999     test   
69726                                        Oldtimer_Lkw  999999     test   
118233  Liebhaber_sucht_Grand_Cherokee_5.9_LX__sofort_...  999999     test   
131785                              Suche_Autos_aller_Art  999999  control   
141004                                  Kaufen_jeder_auto  999999     test   
155418                                   BMW_///_Bodensee  999999  control   
159974                                        Manta_b_gsi  999999  control   
169436                                  KAUFEN_JEDEN_AUTO  999999     test   
169656                                   BMW_///_Bodensee  999999  control   
174617                                    BMW_523I_TAUSCH  999999     test   
301198                 4_Wochen_und_kein_Ergebnis!_Hilfe!  999999  control   
312909                   Ford_Focus_Turnier_1.6_16V_Style  999999     test   
332531  SUCHE_Chevrolet_oder_Gmc_pick_up_vor/bis_bj86_...  999999  control   

       vehicle_type  registration_year    gearbox  power_ps        model  \
38445           bus               1981    manuell        70  transporter   
69726        andere               1960    manuell      9999          NaN   
118233          suv               1998  automatik       241        grand   
131785          NaN               2000        NaN         0          NaN   
141004    limousine               2000        NaN       100          NaN   
155418          NaN               1111        NaN         0          NaN   
159974          NaN               1980        NaN       115       andere   
169436          NaN               2000        NaN         0           a5   
169656          NaN               9999        NaN         0          NaN   
174617        kombi               1997    manuell       174          5er   
301198          NaN               2015        NaN         0          NaN   
312909        kombi               2009    manuell       101        focus   
332531          NaN               1985        NaN         0          NaN   

        odometer_km  registration_month fuel_type           brand  \
38445        150000                   1    benzin      volkswagen   
69726        150000                   0       NaN   mercedes_benz   
118233       150000                   1    benzin            jeep   
131785       100000                   0       NaN  sonstige_autos   
141004       150000                   5    diesel             bmw   
155418         5000                   0       NaN             bmw   
159974        50000                   0       NaN            opel   
169436       100000                   2       NaN            audi   
169656        10000                   0       NaN             bmw   
174617       150000                  11       NaN             bmw   
301198         5000                   2       NaN  sonstige_autos   
312909       125000                   4    benzin            ford   
332531       150000                   0    diesel       chevrolet   

       unrepaired_damage           ad_created  postal_code  \
38445                NaN  2016-03-21 00:00:00        99880   
69726                NaN  2016-03-17 00:00:00        27616   
118233              nein  2016-03-09 00:00:00        76829   
131785               NaN  2016-04-03 00:00:00        21075   
141004               NaN  2016-04-02 00:00:00        10439   
155418               NaN  2016-03-27 00:00:00        88677   
159974               

**Price observations at $999,999**\
After checking the details of the 15 listsings at 999,999 these prices can not be considered serious. 

In [19]:
#Examine some high priced vehicles to determine their viability. 
high_priced = autos.loc[autos['price'] == 195000]
high_priced

date_crawled  \
91514   2016-03-14 19:47:43   
97589   2016-03-21 11:55:05   
356805  2016-03-17 10:48:09   

                                                     name   price   abtest  \
91514   Porsche_911__991__GT3_neuwertig_SportChrono_PD...  195000  control   
97589                            Porsche_Suche_911_R_!!!!  195000  control   
356805                               MERCEDES_SLS_6.3_AMG  195000     test   

       vehicle_type  registration_year    gearbox  power_ps   model  \
91514         coupe               2016  automatik       476     911   
97589         coupe               2016  automatik       500     911   
356805       cabrio               2013  automatik       574  andere   

        odometer_km  registration_month fuel_type          brand  \
91514          5000                   3    benzin        porsche   
97589          5000                   2    benzin        porsche   
356805         5000                   5    benzin  mercedes_benz   

       unrepaired_damage           ad_created  postal_code  \
91514               nein  2016-03-14 00:00:00        44135   
97589               nein  2016-03-21 00:00:00        44143   
356805              nein  2016-03-17 00:00:00        38226   

                  last_seen  
91514   2016-03-29 19:47:00  
97589   2016-03-21 11:55:05  
356805  2016-03-17 10:48:09

In [20]:
high_priced = autos.loc[autos['price'] == 350000]
high_priced

date_crawled                              name   price  \
56490   2016-03-27 18:37:37                       Porsche_991  350000   
108590  2016-03-21 22:37:54                     Porsche_911_R  350000   
120460  2016-03-14 10:44:51                  Andere_ISO_Grifo  350000   
330148  2016-04-04 11:06:24  Mercedes_Benz_C_63_AMG_7G_TRONIC  350000   

         abtest vehicle_type  registration_year    gearbox  power_ps  \
56490   control        coupe               2016    manuell       500   
108590  control          NaN               2015        NaN         0   
120460  control       cabrio               1967    manuell       300   
330148  control    limousine               2009  automatik       457   

           model  odometer_km  registration_month fuel_type           brand  \
56490        911         5000                   3    benzin         porsche   
108590       911         5000                   0       NaN         porsche   
120460       NaN        60000                   5    benzin  sonstige_autos   
330148  c_klasse       100000                   3    benzin   mercedes_benz   

       unrepaired_damage           ad_created  postal_code  \
56490               nein  2016-03-27 00:00:00        70499   
108590               NaN  2016-03-21 00:00:00        55626   
120460              nein  2016-03-14 00:00:00        40217   
330148              nein  2016-04-04 00:00:00        65479   

                  last_seen  
56490   2016-03-27 18:37:37  
108590  2016-03-22 00:42:10  
120460  2016-03-15 23:47:09  
330148  2016-04-04 17:23:22

**Observations at 195K and 350K**\
Due to the cars all being Mercedes or Porsche 911, it is safe to assume that these vehicles have serious prices.

**Price observations for high and low priced vehicles:**\
I believe the vehicles with price <= 350000 are valid. Vehicles above this price, although possible for a new car, or slight used car, is highly unprobable to be sold on eBay. There are also very few vehciles above this number, so if there are a few true entries dropped, it wont be a great loss.  Many prices above 350k are odd numbers, Example 999999, 123456. Therefore we will remove these numbers and replace with NaN.

As for the vehicles that are priced under 100$, it is possible they are listed to be hauled away for free, but more likely someone posted a price that was up for negotiation. Since we are looking for data that would be valued to someone looking to purchase a car, and not haul one away, we should not take any postings with a price below zero seriously and replace with NaN.

In [21]:
autos.loc[(autos['price'] > 350000),'price_$'] = np.nan 
autos.loc[(autos['price'] < 100), 'price_$'] = np.nan
autos['price'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

### 4) Date Range and Date Covers
Moving onto the . There are 5 columns with date values that need to be changed into a correct format. 

In [22]:
print(autos.dtypes)
autos[['date_crawled','ad_created','last_seen', 'registration_month', 'registration_year']][0:5]

date_crawled           object
name                   object
price                   int64
abtest                 object
vehicle_type           object
registration_year       int64
gearbox                object
power_ps                int64
model                  object
odometer_km             int64
registration_month      int64
fuel_type              object
brand                  object
unrepaired_damage      object
ad_created             object
postal_code             int64
last_seen              object
price_$               float64
dtype: object


date_crawled           ad_created            last_seen  \
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57   
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50   
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46   
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17   
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21   

   registration_month  registration_year  
0                   0               1993  
1                   5               2011  
2                   8               2004  
3                   6               2001  
4                   7               2008

**Date Observations:**\
Date_crawled, ad_created, and last_seen are all objects that need to be corrected to numerical format. Calculate the distribution of values in these three columns

In [23]:
#Grab the dates by selecting the first 10 characters
print(autos['date_crawled'].str[:10].head())
print(autos['ad_created'].str[:10].head())
print(autos['last_seen'].str[:10].head())

0    2016-03-24
1    2016-03-24
2    2016-03-14
3    2016-03-17
4    2016-03-31
Name: date_crawled, dtype: object
0    2016-03-24
1    2016-03-24
2    2016-03-14
3    2016-03-17
4    2016-03-31
Name: ad_created, dtype: object
0    2016-04-07
1    2016-04-07
2    2016-04-05
3    2016-03-17
4    2016-04-06
Name: last_seen, dtype: object


In [24]:
(autos['date_crawled']
     .str[:10]
     .value_counts(normalize = True)
     .sort_index() *100
      )

2016-03-05    2.555124
2016-03-06    1.446190
2016-03-07    3.569045
2016-03-08    3.345374
2016-03-09    3.421007
2016-03-10    3.256282
2016-03-11    3.272701
2016-03-12    3.619377
2016-03-13    1.573502
2016-03-14    3.627452
2016-03-15    3.345105
2016-03-16    3.014847
2016-03-17    3.166114
2016-03-18    1.312687
2016-03-19    3.529209
2016-03-20    3.635258
2016-03-21    3.573889
2016-03-22    3.246862
2016-03-23    3.197067
2016-03-24    2.991161
2016-03-25    3.293426
2016-03-26    3.196529
2016-03-27    3.027497
2016-03-28    3.511175
2016-03-29    3.416970
2016-03-30    3.352910
2016-03-31    3.188454
2016-04-01    3.411587
2016-04-02    3.507946
2016-04-03    3.873463
2016-04-04    3.761224
2016-04-05    1.282272
2016-04-06    0.316261
2016-04-07    0.162034
Name: date_crawled, dtype: float64

In [25]:
(autos['date_crawled']
     .str[:10]
     .value_counts(normalize = True, dropna = False)
     .sort_values() *100
      )

2016-04-07    0.162034
2016-04-06    0.316261
2016-04-05    1.282272
2016-03-18    1.312687
2016-03-06    1.446190
2016-03-13    1.573502
2016-03-05    2.555124
2016-03-24    2.991161
2016-03-16    3.014847
2016-03-27    3.027497
2016-03-17    3.166114
2016-03-31    3.188454
2016-03-26    3.196529
2016-03-23    3.197067
2016-03-22    3.246862
2016-03-10    3.256282
2016-03-11    3.272701
2016-03-25    3.293426
2016-03-15    3.345105
2016-03-08    3.345374
2016-03-30    3.352910
2016-04-01    3.411587
2016-03-29    3.416970
2016-03-09    3.421007
2016-04-02    3.507946
2016-03-28    3.511175
2016-03-19    3.529209
2016-03-07    3.569045
2016-03-21    3.573889
2016-03-12    3.619377
2016-03-14    3.627452
2016-03-20    3.635258
2016-04-04    3.761224
2016-04-03    3.873463
Name: date_crawled, dtype: float64

 **Observations:**  There is roughly one month of data, and there are not large spikes on any particular day. Looks like the data was 'crawled' daily.

In [26]:
(autos['last_seen']
     .str[:10]
     .value_counts(normalize = True, dropna = False)
     .sort_values() *100
      )

2016-03-05     0.129196
2016-03-06     0.413428
2016-03-07     0.526205
2016-03-18     0.693084
2016-03-08     0.805592
2016-03-13     0.848927
2016-03-09     0.999386
2016-03-10     1.156306
2016-03-14     1.230055
2016-03-11     1.304612
2016-03-26     1.616029
2016-03-19     1.631371
2016-03-15     1.641061
2016-03-16     1.641868
2016-03-27     1.690855
2016-03-23     1.814937
2016-03-25     1.909681
2016-03-24     1.923677
2016-03-20     1.991505
2016-03-21     2.013576
2016-03-22     2.060679
2016-03-28     2.227288
2016-03-29     2.331184
2016-03-12     2.340066
2016-03-30     2.385554
2016-04-01     2.402242
2016-03-31     2.423774
2016-04-02     2.501561
2016-04-03     2.536552
2016-04-04     2.565352
2016-03-17     2.875961
2016-04-05    12.620583
2016-04-07    12.964837
2016-04-06    21.783015
Name: last_seen, dtype: float64

**Observations last_seen:** Due to the values drastically increasing towards the end of the month, we assume that these cars were not sold during this time, but rather it was just the last time the vehicles were seen due to the end of the crawling period. 

In [27]:
# Find number of different days ads were created
print(autos["ad_created"].str[:10].unique().shape)
        
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )



(114,)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.038893
2016-04-04    0.037741
2016-04-05    0.011655
2016-04-06    0.003155
2016-04-07    0.001558
Name: ad_created, Length: 114, dtype: float64

**Observations** There were 114 different days where ads were created. Some of the ads are very old.  

### 5) Incorrect Registration Year Data
Any registration year above 2016 is innacurate. Its also safe to assume anything before 1900 would be incorrect. 

In [28]:
# How much would we lose if we dropped ads with these registrations years?
(~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0]

0.03969552765874981

Since these ads make up less than 4% of the data, these rows can be removed. 

In [29]:
auto_mask = autos['registration_year'].between(1900,2016)
autos = autos[auto_mask]
autos['registration_year'].value_counts(normalize=True).head(10)*100

2000    6.881271
1999    6.381243
2005    6.254835
2006    5.670161
2001    5.666797
2003    5.570099
2004    5.534503
2002    5.378384
1998    5.031392
2007    4.953473
Name: registration_year, dtype: float64

Most vehicles were registered in the last 20 years. 

### 6) Car sales by brand

In [30]:
autos['brand'].value_counts()

volkswagen        75777
bmw               39137
opel              38200
mercedes_benz     34235
audi              31879
ford              24564
renault           16968
peugeot           10653
fiat               9190
seat               6647
skoda              5497
mazda              5476
smart              5032
citroen            4950
nissan             4841
toyota             4548
sonstige_autos     3799
hyundai            3507
mini               3286
volvo              3257
mitsubishi         2949
honda              2707
kia                2454
alfa_romeo         2265
suzuki             2254
porsche            2194
chevrolet          1789
chrysler           1406
dacia               874
daihatsu            780
jeep                780
land_rover          762
subaru              758
jaguar              614
trabant             578
saab                518
daewoo              513
rover               463
lancia              461
lada                218
Name: brand, dtype: int64

In [39]:
# examine the brand column
print(autos['brand'].value_counts(normalize = True))

volkswagen        0.212391
bmw               0.109695
opel              0.107069
mercedes_benz     0.095955
audi              0.089352
ford              0.068849
renault           0.047559
peugeot           0.029859
fiat              0.025758
seat              0.018631
skoda             0.015407
mazda             0.015348
smart             0.014104
citroen           0.013874
nissan            0.013569
toyota            0.012747
sonstige_autos    0.010648
hyundai           0.009830
mini              0.009210
volvo             0.009129
mitsubishi        0.008266
honda             0.007587
kia               0.006878
alfa_romeo        0.006348
suzuki            0.006318
porsche           0.006149
chevrolet         0.005014
chrysler          0.003941
dacia             0.002450
daihatsu          0.002186
jeep              0.002186
land_rover        0.002136
subaru            0.002125
jaguar            0.001721
trabant           0.001620
saab              0.001452
daewoo            0.001438
r

In [67]:
all_brands = autos['brand'].value_counts(normalize=True)
all_brands = all_brands[:5].index
brand_mean_prices = {} # Empty dict to store brand and its average price. 

for each in all_brands:
    brand_mask = autos[autos['brand'] == each]
    avg_price = brand_mask['price'].mean()
    brand_mean_prices[each] = int(avg_price)
    
# volkswagen = autos[autos['brand'] == 'volkswagen']
# print(volkswagen['price'].mean())

brand_mean_prices

{'volkswagen': 13643,
 'bmw': 14798,
 'opel': 3248,
 'mercedes_benz': 17614,
 'audi': 16218}

**Observation:**  Top five brands are Volkswagen, BMW, Mercedes Benz, Audi, and Opel. Opel is much lower when it comes to the average price of their vehicles. 

### 7) Car Mileage
Goal to look at car mileage across the top 5 brands and see if there is and correlation with price, and maybe find which cars hold their value the best. 
For this we will use pandas series constructor and pandas dataframe constructer

In [77]:
#Create dictionary that holds the brands mean mileage
brand_mean_mileage = {}

for each in all_brands:
    brand_mask = autos[autos['brand'] == each]
    avg_mileage = brand_mask['odometer_km'].mean()
    brand_mean_mileage[each] = int(avg_mileage)
    
brand_mean_mileage

{'volkswagen': 128337,
 'bmw': 132657,
 'opel': 128755,
 'mercedes_benz': 130580,
 'audi': 129491}

In [81]:
#convert the mean price and mean_milage dictionaries to series
brand_mean_prices_series = pd.Series(brand_mean_prices).sort_values()
brand_mean_mileage_series = pd.Series(brand_mean_mileage).sort_values()
print(brand_mean_prices_series)
print(brand_mean_mileage_series)

opel              3248
volkswagen       13643
bmw              14798
audi             16218
mercedes_benz    17614
dtype: int64
volkswagen       128337
opel             128755
audi             129491
mercedes_benz    130580
bmw              132657
dtype: int64


In [82]:
#Create dataframe with the two series
brand_info = pd.DataFrame(brand_mean_prices_series, columns=['mean_price'])
brand_info['mean_mileage'] = brand_mean_mileage_series
brand_info

mean_price  mean_mileage
opel                 3248        128755
volkswagen          13643        128337
bmw                 14798        132657
audi                16218        129491
mercedes_benz       17614        130580

**Oberservation of price and mileage:** Opel does not hold their price well when it comes to high mileage. Why this is the case should be the determining factor if some buys a opel or one of the other 4 top brands. 